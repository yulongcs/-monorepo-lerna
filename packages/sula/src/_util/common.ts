import assignWith from 'lodash/assignWith';
import isUndefined from 'lodash/isUndefined';
import isArray from 'lodash/isArray';

function withDefined(dest: any, src: any) {
  return isUndefined(src) ? dest : src;
}

export function assignWithDefined(...args: Record<string, any>[]) {
  return assignWith(...args, withDefined);
}

export function toArray<T>(value?: T | T[] | null): T[] {
  if (value === undefined || value === null) {
    return [];
  }

  // @ts-ignore
  return isArray(value) ? value : [value];
}


