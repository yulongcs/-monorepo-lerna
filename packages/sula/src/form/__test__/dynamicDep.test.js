import React from 'react';
import { mount } from 'enzyme';
import { Form } from '..';
import { delay } from '../../__tests__/common';

function formMount(props = {}) {
  let formRef;
  const wrapper = mount(
    <Form
      {...props}
      ref={(ref) => {
        formRef = ref;
      }}
      layout="vertical"
      fields={[
        { name: 'username', label: '姓名', field: 'input' },
        {
          name: 'users',
          label: '动态信息',
          isList: true,
          field: {
            type: 'dynamicdepfieldcomp',
            props: {
              fields: [
                {
                  name: 'first',
                  rules: [{ required: true, message: 'Missing first name' }],
                  field: {
                    type: 'input',
                    props: {
                      className: 'input-test',
                    },
                  },
                },
                {
                  name: 'last',
                  rules: [{ required: true, message: 'Missing last name' }],
                  field: 'input',
                  dependency: {
                    value: {
                      relates: ['first'],
                      inputs: [['95']],
                      output: '27',
                      defaultOutput: '98',
                    },
                  },
                },
              ],
            },
          },
        },
      ]}
    />,
  );
  return {
    wrapper,
    formRef,
  };
}

describe('form', () => {
  describe('form  props', () => {
    it('input number', () => {
      const { wrapper } = formMount({});

      let input = wrapper.find('input');
      let button = wrapper.find('button');
      expect(input.length).toEqual(1);
      expect(button.length).toEqual(1);
      expect(wrapper.render()).toMatchSnapshot();

      button.first().simulate('click');
      wrapper.update();

      expect(wrapper.find('input').length > 1).toBeTruthy();
    });

    it('values', async () => {
      const { wrapper, formRef } = formMount({});

      const button = wrapper.find('button');
      button.first().simulate('click');
      expect(formRef.getFieldsValue().users.length).toEqual(1);
      expect(formRef.getFieldsValue().users[0]).toEqual({ first: undefined, last: undefined });
      await delay(0);

      wrapper
        .find('.input-test')
        .at(0)
        .simulate('change', { target: { value: '9' } });

      expect(formRef.getFieldsValue().users[0].last).toEqual('98');

      wrapper
        .find('.input-test')
        .at(0)
        .simulate('change', { target: { value: '95' } });

      expect(formRef.getFieldsValue().users[0].last).toEqual('27');

      wrapper.update();
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('onValuesChange', async () => {
      const onValuesChange = jest.fn();
      const { wrapper, formRef } = formMount({
        onValuesChange,
      });

      wrapper.find('button').last().simulate('click');

      expect(onValuesChange).not.toHaveBeenCalled();
      await delay(0);

      expect(onValuesChange).toHaveBeenCalled();
      expect(onValuesChange).toHaveBeenCalledWith(
        { users: [undefined] },
        { username: undefined, users: [undefined] },
      );

      wrapper.find('button').last().simulate('click');
      await delay(0);

      expect(onValuesChange).toHaveBeenCalledWith(
        { users: [undefined, undefined] },
        { username: undefined, users: [undefined, undefined] },
      );

      wrapper
        .find('.input-test')
        .at(0)
        .simulate('change', { target: { value: '9' } });

      expect(onValuesChange).toHaveBeenCalledWith(
        { users: [{ first: '9' }] },
        { username: undefined, users: [{ first: '9' }, undefined] },
      );

      const delIcon = wrapper.find('.anticon-minus-circle').at(0);

      delIcon.simulate('click');
      await delay(0);

      expect(formRef.getFieldsValue()).toEqual({
        username: undefined,
        users: [{ first: undefined, last: undefined }],
      });

      expect(wrapper.render()).toMatchSnapshot();
    });
  });
});
