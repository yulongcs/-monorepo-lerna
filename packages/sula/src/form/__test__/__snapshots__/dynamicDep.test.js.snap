// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form form  props input number 1`] = `
<form
  class="ant-form ant-form-vertical"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="username"
        title="姓名"
      >
        姓名
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="username"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        title="动态信息"
      >
        动态信息
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <button
            class="ant-btn ant-btn-dashed ant-btn-block"
            style="width: 300px;"
            type="button"
          >
            <span
              aria-label="plus"
              class="anticon anticon-plus"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="plus"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <defs />
                <path
                  d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                />
                <path
                  d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                />
              </svg>
            </span>
            <span>
              添加
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`form form  props onValuesChange 1`] = `
<form
  class="ant-form ant-form-vertical"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="username"
        title="姓名"
      >
        姓名
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="username"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        title="动态信息"
      >
        动态信息
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-space ant-space-horizontal ant-space-align-baseline"
            style="display: flex; margin-bottom: 8px;"
          >
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <div
                class="ant-row ant-form-item"
                dependency="[object Object]"
              >
                <div
                  class="ant-col ant-col-24 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input input-test"
                        id="users_0_first"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <div
                class="ant-row ant-form-item"
                dependency="[object Object]"
              >
                <div
                  class="ant-col ant-col-24 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input"
                        id="users_0_last"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-space-item"
            >
              <span
                aria-label="minus-circle"
                class="anticon anticon-minus-circle"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="minus-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M696 480H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
                  />
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <button
            class="ant-btn ant-btn-dashed ant-btn-block"
            style="width: 300px;"
            type="button"
          >
            <span
              aria-label="plus"
              class="anticon anticon-plus"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="plus"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <defs />
                <path
                  d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                />
                <path
                  d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                />
              </svg>
            </span>
            <span>
              添加
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`form form  props values 1`] = `
<form
  class="ant-form ant-form-vertical"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="username"
        title="姓名"
      >
        姓名
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="username"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        title="动态信息"
      >
        动态信息
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-space ant-space-horizontal ant-space-align-baseline"
            style="display: flex; margin-bottom: 8px;"
          >
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <div
                class="ant-row ant-form-item ant-form-item-is-validating"
                dependency="[object Object]"
              >
                <div
                  class="ant-col ant-col-24 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input input-test"
                        id="users_0_first"
                        type="text"
                        value="95"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <div
                class="ant-row ant-form-item"
                dependency="[object Object]"
              >
                <div
                  class="ant-col ant-col-24 ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        class="ant-input"
                        id="users_0_last"
                        type="text"
                        value="27"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-space-item"
            >
              <span
                aria-label="minus-circle"
                class="anticon anticon-minus-circle"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="minus-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M696 480H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
                  />
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <button
            class="ant-btn ant-btn-dashed ant-btn-block"
            style="width: 300px;"
            type="button"
          >
            <span
              aria-label="plus"
              class="anticon anticon-plus"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="plus"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <defs />
                <path
                  d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                />
                <path
                  d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                />
              </svg>
            </span>
            <span>
              添加
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;
